cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java

```diff
@@ -368,6 +369,20 @@ private int appendEvents(long currentTimeMillis, boolean forced) {
       eventsAppended++;
     }
 
+    // For each partition, if there is no more event in the event queue, update the checkpoint nextOffset
+    for (Int2LongMap.Entry entry : offsets.int2LongEntrySet()) {
+      int partition = entry.getIntKey();
+      if (eventQueue.isEmpty(partition)) {
+        MutableCheckpoint checkpoint = checkpoints.get(partition);
+        long offset = entry.getLongValue();
+        // If the process offset is larger than the offset in the checkpoint and the queue is empty for that partition,
+        // it means everything before the process offset must had been written to the appender.
+        if (checkpoint != null && offset > checkpoint.getNextOffset()) {
```

Comment:
Consider the following are in the queue:

```
event-time   | 108 | 109 | 110 |
next-offset  |  12 |  11 |  10 |
```

As per the logic above, we get the smallest offset before removing the item

After taking `108`, smallest offset is `10`
After taking `109`, smallest offset is `10`
After taking `110`, smallest offset is `10`

Now the queue is empty, technically we want to store `12` in the checkpoint instead of `10`. However, since the queue is emptied, there is no way to get back `12`. But what we can rely on is the offsets being updated from the fetching to be the latest.